### ISSUE TITLE
Helm Service is not correctly replacing keptn variables in generated helm charts

### ISSUE THREAD

**Issue by grabnerandi** [keptn/keptn#2881]
When onboarding a service the Helm service is generating additional helm charts for e.g: canary/primary deployments.

I have observed a weird behavior that leads to incorrectly generated deployment.yamls for the primary deployment.
Here is a snippet of my initial deployment yaml including several usages of the variable .Values.keptn.deployment:

```
spec:
  replicas: {{ .Values.replicaCount }}
  selector:
    matchLabels:
      app: simplenode
  template:
    metadata: 
      labels:
        app: simplenode
        app.kubernetes.io/name: {{ .Values.keptn.service }}
        app.kubernetes.io/part-of: "{{ .Values.keptn.project }}"
        app.kubernetes.io/managed-by: Keptn
        app.kubernetes.io/version: {{ (split ":" .Values.image)._1 | default "latest" }}
    spec:
      containers:
      - name: simplenode
        image: "{{ .Values.image }}"
        imagePullPolicy: IfNotPresent
        env:
        - name: DT_CUSTOM_PROP
          value: "keptn_project={{ .Values.keptn.project }} keptn_service={{ .Values.keptn.service }} keptn_stage={{ .Values.keptn.stage }} keptn_deployment={{ .Values.keptn.deployment }}"
        - name: KEPTN_PROJECT
          value: "{{ .Values.keptn.project }}"
        - name: KEPTN_STAGE
          value: "{{ .Values.keptn.stage }}"
        - name: KEPTN_SERVICE
          value: "{{ .Values.keptn.service }}"
        - name: KEPTN_DEPLOYMENT
          value: "{{ .Values.keptn.deployment }}"
``` 

When using this during `keptn onboard service` the resulting file simplenode-generated/templates/simplenode-primary-deployment.yaml looks like this where the deployment value is correctly replaced in DT_CUSTOM_PROP (shows as primary) but incorrectly in KEPTN_DEPLOYMENT (it is shown as canary)

```
spec:
  replicas: 1
  selector:
    matchLabels:
      app: simplenode-primary
  template:
    metadata:
      labels:
        app: simplenode-primary
        app.kubernetes.io/name: simplenode-primary
        app.kubernetes.io/part-of: student014
        app.kubernetes.io/version: 3.0.0
    spec:
      containers:
      - env:
        - name: DT_CUSTOM_PROP
          value: keptn_project=student014 keptn_service=simplenode keptn_stage=prod
            keptn_deployment=primary
        - name: KEPTN_PROJECT
          value: student014
        - name: KEPTN_STAGE
          value: prod
        - name: KEPTN_SERVICE
          value: simplenode
        - name: KEPTN_DEPLOYMENT
          value: canary
```

I am not sure how it is possible that the same placeholder is replaced with two different values. This currently leads to inconsistent information as this data is passed to the monitoring tool

**Environment**
* Keptn Version (`keptn version`): 0.7.3
* Kubernetes version (`kubectl version`): 01.17

**Affected Component**
* [ ] Docs
* [ ] CLI
* [ ] Kubernetes Integration
* [ ] Openshift Integration
* [x] Helm
* [ ] Istio
* [ ] Bridge
* [ ] Gatekeeper
* [ ] Datastore
* [ ] REST API
* [ ] eventbroker / distributor
* [ ] jmeter

**Comment by bacherfl** [comment]
I just had a look at the code and it seems that this problem occurs because only the string `keptn_deployment=canary` gets replaced by `keptn_deployment_primary` (see [here](https://github.com/keptn/keptn/blob/047a7f7898950941496969d3055671ccf8968347/helm-service/pkg/helm/generated_chart_handler.go#L192)). That's why the other occurences of `{{ .Values.keptn.deployment }} have not been replaced correctly.

**Comment by agrimmer** [comment]
When the `helm-service` is generating the primary Helm chart, it uses the applied manifests of the user Helm chart. Hence,  `{{ .Values.keptn.deployment }}` had been replaced by its value. 
When the `helm-service` now generates the deployments for the primary Helm chart, it searches for the string `keptn_deployment=canary` and replaces it with `keptn_deployment=primary`.

In order to solve this issue, I would propose to restrict the usage of `{{ .Values.keptn.deployment }}` in the following way:
The use of `{{ .Values.keptn.deployment }}` must only be allowed in combination with `keptn_deployment=`. Then the string replacement works again.

@grabnerandi Would that work for your use case?

**Comment by grabnerandi** [comment]
Hi @agrimmer - YES - that would work. I would however then also include a quick "sanity check" when passing the chart during "keptn onboard service". We should notify the user about any potential issues in the provided helm chart

**Comment by christian-kreuzberger-dtx** [comment]
@johannes-b  I believe this has been fixed by @bacherfl in #2883, but the Issue is unassigned, unlabeled and not in a milestone.
